File: android/app/src/main/kotlin/com/leessmin/my_dashcam/activity/camera/CameraActivityOptions.kt
package com.leessmin.my_dashcam.activity.camera

import android.util.Log
import android.view.Surface
import androidx.activity.ComponentActivity
import com.leessmin.my_dashcam.activity.camera.common.FrameRate
import com.leessmin.my_dashcam.activity.camera.common.QualityEnum
import com.leessmin.my_dashcam.activity.camera.data.controller.CameraControllerOptions

/**
 * CameraActivity 配置
 */
data class CameraActivityOptions(
    // 视频质量
    val quality: QualityEnum,
    // 旋转角度, 0=>0度 1=>90度， 2=>180度 3=>270度
    val rotation: Int,
    // 视频帧率
    val frameRate: FrameRate,
    // 是否记录车速
    val recordSpeed: Boolean,
    // 是否记录声音
    val recordVoice: Boolean,
    // 是否启用防抖
    val stabilization: Boolean,
    // 分段录制间隔时间 [1,3,5,10], 单位分钟
    val interval: Int,
    // 录制存储的目录
    val videoSavePath: String,
    // 进入activity是否锁定屏幕
    val lockScreen: Boolean,
    // 最大存储空间 单位GB, 0 代表未开启
    val maxSaveSize: Int,
) {
    companion object {
        private const val TAG = "CameraActivityOptions"

        fun factory(activity: ComponentActivity): CameraActivityOptions {
            val extras = activity.intent.extras
            if (extras == null) {
                Log.e(TAG, "activity.intent.extras为空")
                return CameraActivityOptions(
                    quality = QualityEnum.HD,
                    rotation = Surface.ROTATION_90,
                    frameRate = FrameRate.FPS_30,
                    recordSpeed = true,
                    recordVoice = true,
                    stabilization = true,
                    interval = 3,
                    videoSavePath = "${activity.filesDir}/video",
                    lockScreen = true,
                    maxSaveSize = 0,
                )
            }

            return CameraActivityOptions(
                quality = QualityEnum.fromKey(extras.getString("quality", "HD")),
                rotation = extras.getInt("rotation", Surface.ROTATION_90),
                frameRate = FrameRate.fromFps(extras.getInt("frameRate", 30)),
                recordSpeed = extras.getBoolean("recordSpeed", true),
                recordVoice = extras.getBoolean("recordVoice", true),
                stabilization = extras.getBoolean("stabilization", true),
                interval = extras.getInt("interval", 3),
                videoSavePath = extras.getString("videoSavePath", "${activity.filesDir}/video"),
                lockScreen = extras.getBoolean("lockScreen", true),
                maxSaveSize = extras.getInt("maxSaveSize", 0)
            )
        }
    }

    fun toCameraControllerOptions(): CameraControllerOptions {
        return CameraControllerOptions(
            quality = quality.value,
            rotation = rotation,
            frameRate = frameRate,
            recordSpeed = recordSpeed,
            recordVoice = recordVoice,
            stabilization = stabilization,
            videoSavePath = videoSavePath,
            maxSaveSize = maxSaveSize
        )
    }
}

File: lib/domain/models/camera_activity_options.dart
import 'package:my_dashcam/common/frame_rate.dart';
import 'package:my_dashcam/common/quality.dart';
import 'package:my_dashcam/common/record_interval.dart';
import 'package:my_dashcam/common/rotation.dart';
import 'package:my_dashcam/common/save_size.dart';
import 'package:my_dashcam/configuration/global_configuration.dart';

/// CameraActivity 配置
class CameraActivityOptions {
  const CameraActivityOptions({
    required this.quality,
    required this.rotation,
    required this.frameRate,
    required this.recordSpeed,
    required this.recordVoice,
    required this.stabilization,
    required this.interval,
    required this.videoSavePath,
    required this.lockScreen,
    required this.maxSaveSize,
  });

  // 视频质量
  final Quality quality;

  // 录制旋转的角度
  final Rotation rotation;

  // 视频的帧率
  final FrameRate frameRate;

  // 是否记录车速
  final bool recordSpeed;

  // 是否记录声音
  final bool recordVoice;

  // 是否启用防抖
  final bool stabilization;

  // 分段录制间隔时间 [1,3,5,10], 单位分钟
  final RecordInterval interval;

  // 录制存储的目录
  final String videoSavePath;

  // 进入activity是否锁定屏幕
  final bool lockScreen;

  // 最大存储空间，单位GB, 0不开启
  final SaveSize maxSaveSize;

  Map<String, dynamic> toMap() {
    return {
      "quality": quality.value,
      "rotation": rotation.value,
      "frameRate": frameRate.value,
      "recordSpeed": recordSpeed,
      "recordVoice": recordVoice,
      "stabilization": stabilization,
      "interval": interval.value,
      "videoSavePath": videoSavePath,
      "lockScreen": lockScreen,
      "maxSaveSize": maxSaveSize.value,
    };
  }

  // CameraActivityOption工厂，不需要传递videoSavePath，默认使用“私有目录/files/video”
  static Future<CameraActivityOptions> create({
    required Quality quality,
    required Rotation rotation,
    required FrameRate frameRate,
    required bool recordSpeed,
    required bool recordVoice,
    required bool stabilization,
    required RecordInterval interval,
    required bool lockScreen,
    required SaveSize maxSaveSize,
  }) async {

    return CameraActivityOptions(
      quality: quality,
      rotation: rotation,
      frameRate: frameRate,
      recordSpeed: recordSpeed,
      recordVoice: recordVoice,
      stabilization: stabilization,
      interval: interval,
      lockScreen: lockScreen,
      maxSaveSize: maxSaveSize,
      videoSavePath: await GlobalConfiguration.videoPath,
    );
  }
}

extension CopyWith on CameraActivityOptions {
  CameraActivityOptions copyWith({
    Quality? quality,
    Rotation? rotation,
    FrameRate? frameRate,
    bool? recordSpeed,
    bool? recordVoice,
    bool? stabilization,
    RecordInterval? interval,
    String? videoSavePath,
    bool? lockScreen,
    SaveSize? maxSaveSize,
  }) {
    return CameraActivityOptions(
      quality: quality ?? this.quality,
      rotation: rotation ?? this.rotation,
      frameRate: frameRate ?? this.frameRate,
      recordSpeed: recordSpeed ?? this.recordSpeed,
      recordVoice: recordVoice ?? this.recordVoice,
      stabilization: stabilization ?? this.stabilization,
      interval: interval ?? this.interval,
      videoSavePath: videoSavePath ?? this.videoSavePath,
      lockScreen: lockScreen ?? this.lockScreen,
      maxSaveSize: maxSaveSize ?? this.maxSaveSize,
    );
  }
}
